### PR TITLE
Either instance 'either' method

### PR DESCRIPTION
**Pull request by ferborva**
In reference to the comment made by @OliverJAsh on [#153](https://github.com/ramda/ramda-fantasy/issues/153) I thought it might be a simple addition and interesting one to have on the either prototype.

### CODE REVIEW

File: src/Either.js [hunk 1]
```diff
@@ -127,4 +127,12 @@ Either.Left = function(value) {
 };
 
 
+// Fold methods
+_Right.prototype.fold = _Left.prototype.fold = function eitherFold(leftFn, rightFn) {
```
Comment (buzzdecafe) [inline src/Either.js]: why not just add it to `Either.prototype` and let `Left` and `Right` inherit?

Comment (ferborva) [inline src/Either.js]: Indeed

File: src/Either.js [hunk 2]
```diff
@@ -127,4 +127,12 @@ Either.Left = function(value) {
 };
 
 
+// Fold
+Either.prototype.fold = function eitherFold(leftFn, rightFn) {
+  if (this instanceof _Left) {
+    return leftFn(this.value);
+  }
+  return rightFn(this.value);
+};
```
Comment (buzzdecafe) [inline src/Either.js]: rather than relying on `instanceof`, let's use the `isLeft` property (which sanctuary also uses), and write this as a ternary expression, e.g.

`return (this.isLeft) ? leftFn(this.value) : rightFn(this.value);`

Comment (davidchambers) [inline src/Either.js]: Or even this:

```javascript
return (this.isLeft ? leftFn : rightFn)(this.value);
```

Comment (bijoythomas) [inline src/Either.js]: Why not re-use the static `Either.either` function and further abstract the check for Left or Right?

```
Either.prototype.fold = function(leftFn, rightFn) {
  return Either.either(leftFn, rightFn)(this)
}
```

**Comment by bijoythomas** [comment]
Also, based on the spec here: https://github.com/fantasyland/fantasy-land#foldable, shouldn't the method be called `reduce` (like the Maybe type) instead of fold?

**Comment by ferborva** [comment]
Yep. To stay in line with fantasyland spec it should be called 'reduce'. On the other hand, using Either.either ... don't know what to say to you on that one. Personally I don't see too much use in it to start with as it seems for reasonable to unfold the either from the instance itself rather than using a static method.

**Comment by davidchambers** [comment]
Start with the general type of [`reduce`][1]:

```haskell
reduce :: Foldable f => f a ~> ((b, a) -> b, b) -> b
```

Replace `Foldable f => f` with `Either x`:

```haskell
reduce :: Either x a ~> ((b, a) -> b, b) -> b
```

So, the method must take a `(b, a) -> b` function and a value of type `b`, and return a `b`. Given `Left('XXX')` we do not have a value of type `b` so our only option is to return the `b` provided to us. Given `Right(42)` we have two options: return the given `b`, or apply the provided function to the given `b` and `42` and return the resulting `b`. The latter is far more useful.

The implementation of [`Either#fantasy-land/reduce`][2] looks something like this:

> ```javascript
> Either.prototype['fantasy-land/reduce'] = function(f, x) {
>   return this.isRight ? f(x, this.value) : x;
> };
> ```

Note that `reduce` cannot operate on the value inside a Left. The function being defined in this pull request is something altogether different.


[1]: https://github.com/fantasyland/fantasy-land#reduce-method
[2]: https://github.com/sanctuary-js/sanctuary/blob/v0.12.2/index.js#L2210-L2228

**Comment by ferborva** [comment]
Meaning that adding the .either static method (as .fold or ... whatever) to the Either instance would not make it Foldable ... Can't see this signature mapped on any of the fantasyland specs: Either a b ~> (a -> c) -> (b -> c) -> c.

**Comment by bijoythomas** [comment]
Thanks for the explanation @davidchambers. But then, why is there an Either.either that allows running a leftFn on the value of a Left instance? 

I can see how for a Maybe.Nothing (which is the absence of a value), having reduce return the default makes sense. But a Left instance does not indicate the absence of a value.

**Comment by davidchambers** [comment]
[`either`][1] is a completely valid function, @bijoythomas. It's just a different function from `reduce`. There's no way to make the types line up.


[1]: http://hackage.haskell.org/package/base-4.9.1.0/docs/Prelude.html#v:either

**Comment by davidchambers** [comment]
Would you mind squashing these commits, @ferborva?

**Comment by ferborva** [comment]
@davidchambers No problem :)

**Comment by davidchambers** [comment]
Thank you, @ferborva. The commit message still refers to `fold`. Could you update it?

File: README.md [hunk 3]
```diff
@@ -12,7 +12,7 @@ not undergone thorough testing/use yet.
 
 | Name            | [Setoid][3]  | [Semigroup][4] | [Functor][5] | [Applicative][6] | [Monad][7] | [Foldable][8] | [ChainRec][16] |
 | --------------- | :----------: | :------------: | :----------: | :--------------: | :--------: | :-----------: | :------------: |
-| [Either][9]     |    **✔︎**     |                |     **✔︎**    |      **✔︎**       |   **✔︎**    |               |     **✔︎**      |
+| [Either][9]     |    **✔︎**     |                |     **✔︎**    |      **✔︎**       |   **✔︎**    |              |     **✔︎**      |
```
Comment (davidchambers) [inline README.md]: Please revert this change.

**Comment by ferborva** [comment]
@davidchambers There we go
